### ISSUE TITLE
Only commit dirty files if GPT tries to edit them

### ISSUE THREAD

**Issue by paul-gauthier** [paul-gauthier/aider#200]
> I may end up moving the whole commit system to only committing the files which have been added to the chat. It's not ideal, because GPT could base its changes on other repo files which are dirty which never actually get commited by the user in their current state.

I wonder about a strategy of 
- check status of each file ChatGPT wants to update
- complain if any are dirty or not in git and require user intervention - maybe give the user the options to 
a) add/commit before changing 
b) change anyway 
c) abort

that would mean we could ditch the whole dirty check on startup which takes a while to get used to and give users more flexibility overall

_Originally posted by @joshuavial in https://github.com/paul-gauthier/aider/issues/121#issuecomment-1664701575_

**Comment by paul-gauthier** [comment]
So I have this implemented in a branch. But I actually don't even bother asking the user to make a decision. If GPT tries to edit a dirty file, aider will just commit the dirty changes (with a nice message) before applying and committing GPT's edits.

This seems sensible, low friction and less confusing. You now never get interrupted and asked about pending dirty changes. Aider just commits them for you to keep them distinct from GPT's edits and preserve the ability to cleanly undo GPT's changes. And dirty changes are only ever committed by aider if the file is about to get modified by GPT.

As always, aider still respected the `--dirty-commits` flag. If set, it will apply GPT's edits without first committing dirty files.

**Comment by paul-gauthier** [comment]
This is available in the main branch. You can get it by installing the latest version from github:

```
python -m pip install git+https://github.com/paul-gauthier/aider.git
```

If you have a chance to try it, let me know if it works better for you.

**Comment by paul-gauthier** [comment]
This shipped in v0.13.0. I'm going to close this issue for now, but feel free to re-open or file a new issue any time.
